cellfinder_napari/detect_utils.py

```diff
@@ -0,0 +1,186 @@
+from dataclasses import dataclass, asdict
+from pathlib import Path
+
+import napari
+from cellfinder_core.main import main as cellfinder_run
+from imlib.cells.cells import Cell
+from napari.qt.threading import thread_worker
+
+from cellfinder_napari.utils import cells_to_array
+
+
+@dataclass
+class InputContainer:
+    """Base data class to group inputs into a container, with persistent defaults"""
+
+    persistent_defaults = (
+        {}
+    )  # some elements need to be remembered without having to create a new instance, for reset button
+
+    @classmethod
+    def numerical_widget(
+        cls, key, custom_label=None, step=0.1, min=None, max=None
+    ) -> dict:
+        """Provide info to populate a numerical widget with the class's persistent defaults."""
+        # TODO: maybe some assertions like either both or neither min/max==None and cls.defaults[key] is numerical.
+        if custom_label is None:
+            label = key.replace("_", " ").capitalize()
+        else:
+            label = custom_label
+        if min is None and max is None:
+            return dict(
+                value=cls.persistent_defaults[key], label=label, step=step
+            )
+        else:
+            return dict(
+                value=cls.persistent_defaults[key],
+                label=label,
+                step=step,
+                min=min,
+                max=max,
+            )
+
+    def as_core_arguments(self) -> dict:
+        """Allow customisation of how inputs are passed to cellfinder-core"""
+        return asdict(self)
+
+
+@dataclass
+class DataInputs(InputContainer):
+    """Container for image-related ("Data") inputs."""
+
```

Comment:
this might be your design preference but I find the first blank line a bit odd 🤷

Reply:
Wasn't there on purpose 😅 Hope I've removed occurrences of this now (and otherwise can be part of formatting PR).